cloud/docker/docker_login.py

```diff
@@ -0,0 +1,251 @@
+            self.module.fail_json(msg="Docker API Error: %s" % e.explanation)
+        except Exception as e:
+            self.module.fail_json(msg="failed to login to the remote registry", error=repr(e))
+
+        # Get status from registry response.
+        if self.response.has_key("Status"):
+            self.log.append(self.response["Status"])
+
+        # Update the dockercfg if not in check mode.
+        if not self.module.check_mode:
+            self.update_dockercfg()
+
+    # This is what the underlaying docker-py unfortunately doesn't do (yet).
+    def update_dockercfg(self):
+
+        # Create dockercfg file if it does not exist.
+        if not os.path.exists(self.dockercfg_path):
+            dockercfg_path_dir = os.path.dirname(self.dockercfg_path)
+            if not os.path.exists(dockercfg_path_dir):
+                os.makedirs(dockercfg_path_dir)
+            open(self.dockercfg_path, "w")
+            self.log.append("Created new Docker config file at %s" % self.dockercfg_path)
+        else:
+            self.log.append("Found existing Docker config file at %s" % self.dockercfg_path)
+
+        # Build a dict for the existing dockercfg.
+        try:
+            docker_config = json.load(open(self.dockercfg_path, "r"))
+        except ValueError:
+            docker_config = dict()
+        if not docker_config.has_key("auths"):
+            docker_config["auths"] = dict()
+        if not docker_config["auths"].has_key(self.registry):
+            docker_config["auths"][self.registry] = dict()
+
+        # Calculate docker credentials based on current parameters.
+        new_docker_config = dict(
+            auth  = base64.b64encode(self.username + b':' + self.password),
+            email = self.email
+        )
+
+        # Update config if persisted credentials differ from current credentials.
+        if new_docker_config != docker_config["auths"][self.registry]:
+            docker_config["auths"][self.registry] = new_docker_config
+            try:
+                json.dump(docker_config, open(self.dockercfg_path, "w"), indent=4, sort_keys=True)
+            except Exception as e:
+                self.module.fail_json(msg="failed to write auth details to file", error=repr(e))
+            self.log.append("Updated Docker config with new credentials.")
+            self.changed = True
+
+    # Compatible to docker-py auth.decode_docker_auth()
+    def encode_docker_auth(self, auth):
+        s = base64.b64decode(auth)
+        login, pwd = s.split(b':', 1)
+        return login.decode('ascii'), pwd.decode('ascii')
+
+    def get_msg(self):
+        return ". ".join(self.log)
+
+    def has_changed(self):
+        return self.changed
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            registry        = dict(required=False, default='https://index.docker.io/v1/'),
+            username        = dict(required=True),
+            password        = dict(required=True),
```

Comment:
Noticed one more thing:  we should mark this as no_log like this:
``` python
    password    = dict(required=True, no_log=True),
```

That way ansible will apply some heuristics to keep that parameter from being displayed in output, logged, etc.